File: 6. Graph/13_Dijkstra_algo.cpp
#include <bits/stdc++.h>
using namespace std;

vector<int> dijkstra(int v, vector<vector<int>> adj[], int src)
{
    vector<int> cost(v, 0);
    cost[src] = 0;

    vector<bool> visited(v, false);
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
    pq.push({0, src});
    while (!pq.empty())
    {
        pair<int, int> p = pq.top();
        int currcost = p.first;
        int currnode = p.second;
        pq.pop();

        if (visited[currnode])
            continue;

        visited[currnode] = true;
        visited[currnode] = currcost;

        for (int i = 0; i < adj[currnode].size(); i++)
        {
            int neighbour = adj[currnode][i][0];
            int weight = adj[currnode][i][1];
            if (visited[neighbour])
                continue;

            pq.push({currcost + weight, neighbour});
        }
    }
    return cost;
}

File: DSA/6. Graph/13_Dijkstra_algo.java
/*
    Problem Link: https://www.geeksforgeeks.org/dijkstras-shortest-path-algorithm-greedy-algo-7/
    
    Dijkstra's Shortest Path Algorithm
*/

import java.util.*;

class Dijkstra_algo {
    public static int[] dijkstra(int v, List<int[]>[] adj, int src) {
        int[] cost = new int[v];
        Arrays.fill(cost, Integer.MAX_VALUE);
        cost[src] = 0;

        boolean[] visited = new boolean[v];
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        pq.offer(new int[]{0, src});
        
        while (!pq.isEmpty()) {
            int[] p = pq.poll();
            int currcost = p[0];
            int currnode = p[1];

            if (visited[currnode]) {
                continue;
            }

            visited[currnode] = true;
            cost[currnode] = currcost; // Fixed: was setting visited[currnode] = currcost

            for (int i = 0; i < adj[currnode].size(); i++) {
                int neighbour = adj[currnode].get(i)[0];
                int weight = adj[currnode].get(i)[1];
                if (visited[neighbour]) {
                    continue;
                }

                pq.offer(new int[]{currcost + weight, neighbour});
            }
        }
        return cost;
    }
}
